### PR TITLE
Collection expressions: support builder methods returning derived or implementing types

### PR DESCRIPTION
**Pull request by cston**
See [*create methods*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#create-methods):

> * There is an [*identity conversion*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/conversions.md#1022-identity-conversion), [*implicit reference conversion*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/conversions.md#1028-implicit-reference-conversions), or [*boxing conversion*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/conversions.md#1029-boxing-conversions) from the method return type to the *collection type*.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 1]
```diff
@@ -855,36 +861,38 @@ internal bool TryGetCollectionIterationType(ExpressionSyntax syntax, TypeSymbol
...
-                    {
-                        continue;
-                    }
+                    methodWithTargetTypeParameters = method;
                 }
 
-                if (!targetType.Equals(method.ReturnType, TypeCompareKind.AllIgnoreOptions))
+                var spanTypeArg = ((NamedTypeSymbol)methodWithTargetTypeParameters.Parameters[0].Type).TypeArgumentsWithAnnotationsNoUseSiteDiagnostics[0].Type;
+                var conversion = Conversions.ClassifyImplicitConversionFromType(elementTypeOriginalDefinition, spanTypeArg, ref candidateUseSiteInfo);
+                if (!conversion.IsIdentity)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:878]: I'm much happier with this logic, thanks Chuck!

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 2]
```diff
@@ -1892,6 +1892,8 @@
     <Field Name="CollectionCreation" Type="BoundExpression?" Null="allow" SkipInVisitor="true"/>
     <!-- Construct method for collection types with [CollectionBuilder]. -->
     <Field Name="CollectionBuilderMethod" Type="MethodSymbol?" Null="allow"/>
+    <!-- Conversion from construct method return type to collection type. -->
+    <Field Name="CollectionBuilderReturnTypeConversion" Type="Conversion"/>
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:1896]: Aleksey previously removed all `Conversion` elements from the bound tree, except for `BoundConversion` itself, and then had each node hold onto an actual `BoundConversion` and a placeholder for the thing being converted, if it couldn't be directly represented. I think we should do the same here: I'd have to go dig up the changes to remember the exact reasons, but I do remember that they were related to some correctness issues.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 3]
```diff
@@ -9498,6 +9498,642 @@ static void Main()
...
+                        return new MyCollection(new List<int>(items.ToArray()));
+                    }
+                }
+                class Program
+                {
+                    static void Main()
+                    {
+                        MyCollectionBase x = [];
+                        x.Report();
+                        MyCollectionBase y = [1, 2, 3];
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:9534]: Can you add a verison of this where the target type is `MyCollection`?

**Review by 333fred — COMMENTED** [review]
Done review pass

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:1896]: Not blocking the PR: Good catch. Could we add something in the BoundNodes generator to catch this in the future?

Comment (333fred) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:1896]: We probably should.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 4]
```diff
@@ -9498,6 +9498,700 @@ static void Main()
...
+                        MyCollection x = [];
+                        MyCollection y = [1, 2, 3];
+                    }
+                }
+                """;
+            var comp = CreateCompilation(
+                new[] { sourceA, sourceB2 },
+                targetFramework: TargetFramework.Net80);
+            comp.VerifyEmitDiagnostics(
+                // 1.cs(5,26): error CS7036: There is no argument given that corresponds to the required parameter 'list' of 'MyCollection.MyCollection(List<int>)'
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:9561]: Do you think it's worth testing a variant of `CollectionBuilderAttribute` that is marked `Inherited = true`?

**Review by 333fred — APPROVED** [review]
Generally LGTM (commit 2). One small question.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:9561]: There is an existing test, `CollectionBuilder_InheritedAttributeOnBaseCollection`. Let me know if that doesn't cover the expected scenario.

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml (hunk 2 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: https://github.com/dotnet/roslyn/pull/69785 will validate this.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 4 above)
Comment (333fred) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:9561]: Thanks, that covers it for me.

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**